### PR TITLE
Parse, validate and use new zone endpoint syntax

### PR DESCRIPTION
**Pull request by jonmv**
@mpolden please review.

### CODE REVIEW

**Comment by jonmv** [comment]
Remaining:

1. Validation override required to remove private endpoint.
2. Removal of old code to parse these settings.
3. Not provisioning a public endpoint, but that's probably more work, as we assume several places this is available.

File: vespa-feed-client/src/test/java/ai/vespa/feed/client/impl/ApacheClusterTest.java [hunk 1]
```diff
@@ -48,9 +48,9 @@ void testClient() throws IOException, ExecutionException, InterruptedException,
                                                  Map.of("name1", () -> "value1",
                                                         "name2", () -> "value2"),
                                                  "content".getBytes(UTF_8),
-                                                 Duration.ofSeconds(1)),
+                                                 Duration.ofSeconds(10)),
```
Comment (jonmv) [inline vespa-feed-client/src/test/java/ai/vespa/feed/client/impl/ApacheClusterTest.java:51]: Ah, yes, not really related, but let's just leave it here. @hakonhall FYI :)

**Comment by jonmv** [comment]
Well ... 2. is just to remove the grammar from services..rnc :)

File: config-model-api/src/main/java/com/yahoo/config/application/api/DeploymentInstanceSpec.java [hunk 2]
```diff
@@ -251,6 +261,16 @@ public boolean deploysTo(Environment environment, RegionName region) {
         return zones().stream().anyMatch(zone -> zone.concerns(environment, Optional.of(region)));
     }
 
+    /** Returns the zone endpoint specified for the given region, or the default, or {@code null}. */
```
Comment (mpolden) [inline config-model-api/src/main/java/com/yahoo/config/application/api/DeploymentInstanceSpec.java:264]: Should be "or empty".

File: config-model-api/src/main/java/com/yahoo/config/application/api/xml/DeploymentSpecXmlReader.java [hunk 3]
```diff
@@ -345,33 +362,132 @@ private List<Endpoint> readEndpoints(Element parent, Optional<String> instance,
...
+                }
+
+                if (zoneEndpointType.isPresent()) {
+                    if (regions.isEmpty()) regions.add(null);
+                    ZoneEndpoint endpoint;
+                    Optional<String> enabled = XML.attribute("enabled", endpointElement);
+                    if ("zone".equals(zoneEndpointType.get())) {
+                        switch (enabled.orElse("true")) {
+                            case "true" -> endpoint = new ZoneEndpoint(true, false, List.of());
+                            case "false" -> endpoint = new ZoneEndpoint(false, false, List.of());
```
Comment (mpolden) [inline config-model-api/src/main/java/com/yahoo/config/application/api/xml/DeploymentSpecXmlReader.java:379]: Nit: Using switch expression to assign `endpoint` would be neater (`endpoint = switch ...`).

File: config-model-api/src/main/java/com/yahoo/config/application/api/xml/DeploymentSpecXmlReader.java [hunk 4]
```diff
@@ -345,33 +362,132 @@ private List<Endpoint> readEndpoints(Element parent, Optional<String> instance,
...
+        endpoints.addAll(endpointsById.values());
+        validateAndConsolidate(endpointsByZone, zoneEndpoints);
+    }
+
+    static void validateAndConsolidate(Map<String, Map<RegionName, List<ZoneEndpoint>>> in, Map<ClusterSpec.Id, Map<ZoneId, ZoneEndpoint>> out) {
+        in.forEach((cluster, regions) -> {
+            List<ZoneEndpoint> wildcards = regions.remove(null);
+            ZoneEndpoint wildcardZoneEndpoint = null;
+            ZoneEndpoint wildcardPrivateEndpoint = null;
+            if (wildcards != null) for (ZoneEndpoint endpoint : wildcards) {
```
Comment (mpolden) [inline config-model-api/src/main/java/com/yahoo/config/application/api/xml/DeploymentSpecXmlReader.java:433]: Missing newline after `if`.

File: config-model-api/src/main/java/com/yahoo/config/application/api/xml/DeploymentSpecXmlReader.java [hunk 5]
```diff
@@ -345,33 +362,132 @@ private List<Endpoint> readEndpoints(Element parent, Optional<String> instance,
...
+                            case "false" -> endpoint = new ZoneEndpoint(false, false, List.of());
+                            default -> throw new IllegalArgumentException("Zone endpoint for container-id '" + containerId + "': " +
+                                                                          "invalid 'enabled' value; must be 'true' or 'false'");
+                        }
+                    }
+                    else {
+                        if (enabled.isPresent()) illegal("Private endpoint for container-id '" + containerId + "': " +
+                                                         "only endpoints of type 'zone' can specify 'enabled'");
+                        List<AllowedUrn> allowedUrns = new ArrayList<>();
+                        for (var allow : XML.getChildren(endpointElement, "allow")) {
```
Comment (mpolden) [inline config-model-api/src/main/java/com/yahoo/config/application/api/xml/DeploymentSpecXmlReader.java:388]: Attempting to set `<allow>` on a non-private endpoints should throw.

File: config-model-api/src/main/java/com/yahoo/config/application/api/xml/DeploymentSpecXmlReader.java [hunk 6]
```diff
@@ -345,33 +362,132 @@ private List<Endpoint> readEndpoints(Element parent, Optional<String> instance,
...
+                    if ("zone".equals(zoneEndpointType.get())) {
+                        switch (enabled.orElse("true")) {
+                            case "true" -> endpoint = new ZoneEndpoint(true, false, List.of());
+                            case "false" -> endpoint = new ZoneEndpoint(false, false, List.of());
+                            default -> throw new IllegalArgumentException("Zone endpoint for container-id '" + containerId + "': " +
+                                                                          "invalid 'enabled' value; must be 'true' or 'false'");
+                        }
+                    }
+                    else {
+                        if (enabled.isPresent()) illegal("Private endpoint for container-id '" + containerId + "': " +
```
Comment (mpolden) [inline config-model-api/src/main/java/com/yahoo/config/application/api/xml/DeploymentSpecXmlReader.java:385]: Should also throw for other types? Looks like `enabled` on global and application endpoints is just ignored now.

File: config-model-api/src/main/java/com/yahoo/config/application/api/DeploymentInstanceSpec.java (hunk 2 above)
Comment (jonmv) [reply config-model-api/src/main/java/com/yahoo/config/application/api/DeploymentInstanceSpec.java:264]: Yes, thanks, forgot doc when I changed the method.

File: config-model-api/src/main/java/com/yahoo/config/application/api/xml/DeploymentSpecXmlReader.java (hunk 3 above)
Comment (jonmv) [reply config-model-api/src/main/java/com/yahoo/config/application/api/xml/DeploymentSpecXmlReader.java:379]: Yes, moved extraction of more data up, and did this instead. Thanks.

File: config-model-api/src/main/java/com/yahoo/config/application/api/xml/DeploymentSpecXmlReader.java (hunk 4 above)
Comment (jonmv) [reply config-model-api/src/main/java/com/yahoo/config/application/api/xml/DeploymentSpecXmlReader.java:433]: Fixed.

File: config-model-api/src/main/java/com/yahoo/config/application/api/xml/DeploymentSpecXmlReader.java (hunk 5 above)
Comment (jonmv) [reply config-model-api/src/main/java/com/yahoo/config/application/api/xml/DeploymentSpecXmlReader.java:388]: Fixed, see above.

File: config-model-api/src/main/java/com/yahoo/config/application/api/xml/DeploymentSpecXmlReader.java (hunk 6 above)
Comment (jonmv) [reply config-model-api/src/main/java/com/yahoo/config/application/api/xml/DeploymentSpecXmlReader.java:385]: Yes, fixed; see above.

**Review by mpolden — APPROVED** [review]
Great! But build is failing, something is still specifying `<load-balancer>`.

**Comment by jonmv** [comment]
Thanks. I wonder what. I searched for that string ... :>

**Comment by jonmv** [comment]
Evidently a poor attempt. Yes, there are some test resources. Thanks again.

**Comment by jonmv** [comment]
I did run the tests, though. Even weirder.

**Comment by jonmv** [comment]
Aha, I didn't intend to push that commit with the removal of old schema grammar. Well, shouldn't be in use anywhere ...

**Comment by jonmv** [comment]
Added some more validation, and fixed some conflicts, PTAL :)

**Review by mpolden — APPROVED** [review]

**Pull request merged**